### ISSUE TITLE
Automatic retry mechanism if task fail to fetch material

### ISSUE THREAD

**Issue by ariefrahmansyah** [gocd/gocd#6252]
##### Issue Type

- Feature proposal

##### Summary

I'm using GitLab as main Git repository and sometimes I have connectivity issue that made git fetching failed. Can GoCD have an automatic retry mechanism to re-trigger the material fetching task?

##### Any other info

I'm thinking for something like Concourse's attempt: https://concourse-ci.org/attempts-step-modifier.html

**Comment by arvindsv** [comment]
@ariefrahmansyah If it's an auto polled material, it should retry in around a minute. If not, there's currently no mechanism to retry (a forced trigger or timer, for instance).

**Comment by ariefrahmansyah** [comment]
@arvindsv sorry, it's not auto poll material, but it's when running task failed to fetch material. the task will be failed.

![image](https://user-images.githubusercontent.com/8122852/57267934-7fad5180-70b4-11e9-8111-57708f030ff6.png)

**Comment by dominik-ba** [comment]
I would really love to have this feature!

**Comment by khandelwaly** [comment]
@chadlwilson is any working on this feature?

**Comment by mattgauntseo-sentry** [comment]
We run into this issue every now and then, it would be great to have a method to set retry count and / or a wait between retries.

**Comment by chadlwilson** [comment]
@mattgauntseo-sentry Have you determined root cause of the issue in your case? Is it a git material or some other material type or plugin material? Is it a temporary connectivity problem or target material source rate limit?

**Comment by mattgauntseo-sentry** [comment]
We use github and it presents itself as a ssh connection issue. It's always intermittent. I don't believe this is a rate limit, but wouldn't rule it out.

I'll share an example log when it happens again.

**Comment by chadlwilson** [comment]
Hmm. GitHub definitely rate limits on https and GoCD's polling nature can be a problem for that, but not 100% sure what it looks like for ssh, and whether it's different on enterprise plans.

Regardless useful to know, as there are probably multiple layers this could be addressed (specific material type, more abstract for all materials - that type of thing.

Is this a standard OOTB GoCD git materials or some extra plugin (eg for PRs)

**Comment by mattgauntseo-sentry** [comment]
This is a standard OOTB git material with an ssh url (git@github.com/...). I believe these are repos in an enterprise plan.

**Comment by mattgauntseo-sentry** [comment]
Just ran into this during a test:

```
[go] Start to prepare <pipeline name>/52/deploy_primary/1/deploy on <elastic agent name>-<elastic agent ID> [/go]
[go] Start to update materials.

[go] Start updating <pipeline group> at revision <commit revision> from git@github.com:<org>/<repo>.git
STDERR: Cloning into '/go/pipelines/<pipeline name>/<pipeline group>'...
[GIT] Fetching changes
STDERR: ssh: connect to host github.com port 22: Connection timed out
STDERR: fatal: Could not read from remote repository.
STDERR: 
STDERR: Please make sure you have the correct access rights
STDERR: and the repository exists.
git fetch failed for [git@github.com:<org>/<repo>.git]
```

**Comment by chadlwilson** [comment]
Yeah, that looks like a general networking issue where at best anything done at GoCD level will be a bit of a fudge. But perhaps a necessary one in some setups.

If you see these types of errors exclusively on agents rather than the server logs you might want to explore anything different about the agent environment or specific to elastic agents (although the server can typically recover due to its polling nature, assuming you're not using webhooks).

**Comment by mattgauntseo-sentry** [comment]
I've seen this on the server too with config repos hitting this error, but in those cases the polling gets us to an eventual good state.
